src/pdmat.jl

```diff
@@ -1,23 +1,49 @@
 """
 Full positive definite matrix together with a Cholesky factorization object.
 """
-struct PDMat{T<:Real,S<:AbstractMatrix} <: AbstractPDMat{T}
+struct PDMat{T<:Real,S<:AbstractMatrix{T}} <: AbstractPDMat{T}
     mat::S
     chol::Cholesky{T,S}
 
-    PDMat{T,S}(m::AbstractMatrix{T},c::Cholesky{T,S}) where {T,S} = new{T,S}(m,c)
+    function PDMat{T,S}(m::AbstractMatrix, c::Cholesky) where {T,S}
+        d = LinearAlgebra.checksquare(m)
+        if size(c, 1) != d
+            throw(DimensionMismatch("Dimensions of mat and chol are inconsistent."))
+        end
+        # in principle we might want to check that `c` is a Cholesky factorization of `m`,
+        # but that's slow
+        return new{T,S}(m,c)
+    end
 end
+PDMat{T,S}(pdm::PDMat) where {T,S} = PDMat{T,S}(pdm.mat, pdm.chol)
 
-function PDMat(mat::AbstractMatrix,chol::Cholesky{T,S}) where {T,S}
-    d = LinearAlgebra.checksquare(mat)
-    if size(chol, 1) != d
-        throw(DimensionMismatch("Dimensions of mat and chol are inconsistent."))
-    end
-    PDMat{T,S}(convert(S, mat), chol)
+function PDMat{T}(m::AbstractMatrix, c::Cholesky) where T
+    c = Cholesky{T}(c)
+    return PDMat{T,mattype(c)}(m, c)
 end
+PDMat{T}(pdm::PDMat) where T = PDMat{T}(pdm.mat, pdm.chol)
 
+PDMat(mat::AbstractMatrix,chol::Cholesky{T,S}) where {T,S} = PDMat{T,S}(mat, chol)
+
+function PDMat{T,S}(mat::AbstractMatrix) where {T,S}
+    mat = convert(S, mat)
+    return PDMat{T,S}(mat, cholesky(mat))
+end
+function PDMat{T}(mat::AbstractMatrix) where T
+    mat = convert(AbstractMatrix{T}, mat)
+    return PDMat{T}(mat, cholesky(mat))
+end
 PDMat(mat::AbstractMatrix) = PDMat(mat, cholesky(mat))
-PDMat(fac::Cholesky) = PDMat(AbstractMatrix(fac), fac)
+
+function PDMat{T,S}(c::Cholesky) where {T,S}
+    c = Cholesky{T,S}(c)
```

Comment:
Similarly, is this guaranteed to be a no-op if `c isa Cholesky{T,S}`? Or should we use
```suggestion
    c = convert(Cholesky{T,S}, c)
```
?